STYLEGUIDE.md

```diff
@@ -0,0 +1,740 @@
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use single quotes */
+const foo = 'foo';
+
+/* Use double quotes if the quoted string contains single quotes. */
+const foo = "'foo'";
+
+/* Use template string if the quoted string contains double and single quotes. */
+const foo = `They know when to use 'single quotes' and "double quotes"`;
+### String concatenation
+When possible use template literals, this makes the code more readable in most cases.
+Feel free to concatenate strings with `+` in case it makes to code better readable.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't concatenate strings with `+` */
+const foo = 'foo ' + bar + ' bar';
+
+/* Gray area: If template literals become too complex, don't use them. */
+const foo = `foo ${fn(`${bar}bar`)}`;
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Concatenate strings using template literals */
+const foo = `foo ${bar} bar`;
+
+/* Gray area: If template literals become too complex, feel free to concatenate with +. */
+const foo = `foo ${fn(bar + 'bar')}`;
+```
+
+## Arrays
+
+### Use `.forEach` rather than `for`
+Use the functional language concepts of JavaScript in favor of the imperative ones, when possible.
+Use the correct function for your use case.
+In rare cases, when performance is of the essence, it makes sense to resort to the imperative `for` instead.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use for ... of if you can use .forEach */
+for (const item of array) {
+  ...
+}
+
+/* Don't use .forEach if you can use .map */
+const result = [];
+array.forEach(item => {
+  result.push(item.property);
+});
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use .forEach if possible */
+array.forEach(item => {
+  ...
+});
+
+/* Use .map, if appropriate */
+const result = array.map(item => item.property);
+```
+
+
+## `null` and `undefined`
+
+### Use optional parameters for optional values
+Use optional parameters, if there is a default value or if it semantically makes sense to omit the parameters when consuming the API.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use explicit undefined if the parameter is semantically optional */
+function foo(options: Options | undefined) {
+  ...
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use optional parameter if the parameter is semantically optional */
+function foo(options?: Options) {
+  ...
+}
+
+/* Use default parameter if there is a default (making the parameter inherently optional for the consumer) */
+function foo(options: Options = {}) {
+  ...
+}
+```
+
+### Use `undefined` for explicit absence of values
+If a value semantically is not optional prefer using `undefined` for values that do not exist or are not defined.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use null for non-existent values */
+function foo(obj: SomeType | null) {
+  ...
+}
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use undefined for non-existent values */
+function foo(obj: SomeType | undefined) {
```

Comment:
why this is better than `function foo(obj?: SomeType) {`

Reply:
There is a section on optional values before. This section is about values that are not optional. The difference is that with optional parameters you can invoke the function like this `foo()`. If the parameter is not semantically optional this doesn't make sense and should not be possible.
Consider: 
```
function getFooFromBar(bar?: Bar) {
  return bar?.foo;
}

// This does not make sense
getFooFromBar();

// But this does
const bar = getBarFromSomewherePotentiallyUndefined();
getFooFromBar(bar);
```

I improved the description a bit, let me know if this is clearer or if we should improve the examples.